js/ui/status/keyboard.js

```diff
@@ -480,6 +483,53 @@ const InputSourceManager = new Lang.Class({
             this._updateMruSettings();
     },
 
+    get passwordModeEnabled() {
+        return this._passwordModeEnabled;
+    },
+
+    set passwordModeEnabled(enable) {
+        this._passwordModeEnabled = enable;
+        this._switchToLatinLayout();
+    },
+
+    _switchToLatinLayout: function() {
+        let source;
+
+        if (this._passwordModeEnabled) {
+            // do not switch the keyboard layout if the current one is a latin one
+            if (this._keyboardManager.isLatinLayout(this._currentSource.id))
+                return;
+
+            // look for latin keyboard layout to switch to
+            for (let i = 0; i < this._mruSources.length; ++i) {
+                let src = this._mruSources[i];
+                if (this._keyboardManager.isLatinLayout(src.id)) {
+                    source = src;
+                    break;
+                }
+            }
+            if (!source) {
+                // fallback to an us keyboard layout
+                source = new InputSource('xkb',
+                                         'us',
+                                         'English (US)',
+                                         'en',
+                                         Object.keys(this._inputSources).length);
+                this._fallbackInputSource = source;
+                this._inputSourcesChanged();
+            }
+            this._lastSource = this._currentSource;
+        } else {
+            if (!this._lastSource)
+                return;
+            source = this._lastSource;
+            this._lastSource = null;
+            this._fallbackInputSource = null;
+            this._inputSourcesChanged();
```

Comment:
Since this is common code, I'd move it to the end.

Reply:
Hmm, the this._inputSourcesChanged(); should not happen in all the cases.

Reply:
I am suggesting that you put it right before the call to `this._activateInputSource()`. That should be equivalent to the current implementation.

Reply:
Doing that would also call "this._inputSourcesChanged()" for the case where we have a latin layout already available and we can switch to it. It should only be called when we add the "en" layout or remove it.